File: src/ros2_bridge/mappers/tf2_msgs/tf_message.rs
//! Typed mapper for `tf2_msgs/msg/TFMessage`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn tf_message_to_bus(msg: ros_env::tf2_msgs::msg::TFMessage) -> crate::tf2_msgs::msg::v1::TfMessage {
    crate::tf2_msgs::msg::v1::TfMessage {
        transforms: msg.transforms.into_iter().map(crate::ros2_bridge::mappers::geometry_msgs::transform_stamped::transform_stamped_to_bus).collect(),
    }
}

pub(crate) fn tf_message_to_ros(bus: crate::tf2_msgs::msg::v1::TfMessage) -> ros_env::tf2_msgs::msg::TFMessage {
    ros_env::tf2_msgs::msg::TFMessage {
        transforms: bus.transforms.into_iter().map(crate::ros2_bridge::mappers::geometry_msgs::transform_stamped::transform_stamped_to_ros).collect(),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct Tf2MsgsTfMessageMapper;

impl TypedTopicMapper for Tf2MsgsTfMessageMapper {
    type Ros = ros_env::tf2_msgs::msg::TFMessage;
    type Bus = crate::tf2_msgs::msg::v1::TfMessage;

    fn type_name(&self) -> &'static str {
        "tf2_msgs/msg/TFMessage"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(tf_message_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(tf_message_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/tf2_msgs/tf_message.py
"""Generated mapper for `tf2_msgs/msg/TFMessage`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.geometry_msgs.transform_stamped import transform_stamped_to_bus, transform_stamped_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.tf2_msgs.msg.v1 import TFMessage as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def tf_message_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.transforms.extend([transform_stamped_to_bus(x) for x in msg.transforms])
    return bus


def tf_message_to_ros(bus):
    from tf2_msgs.msg import TFMessage as RosMsg

    out = RosMsg()
    out.transforms = [transform_stamped_to_ros(x) for x in bus.transforms]
    return out


class Tf2MsgsTfMessageMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from tf2_msgs.msg import TFMessage as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return tf_message_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return tf_message_to_ros(bus)
